File: Domain/Model/InterviewMessage.swift
import Foundation

/// Audio source for speaker identification
enum AudioSource {
    case microphone    // Your voice (you speaking)
    case systemAudio   // Interviewer voice (Zoom/Teams output)
}

/// Represents a message in the interview timeline
struct InterviewMessage: Identifiable {
    let id: UUID
    let timestamp: Date
    let type: MessageType
    let content: String
    let topic: String?
    var screenshotId: UUID?  // For screenshot type messages
    var audioSource: AudioSource?  // Source of audio for speaker identification
    var isCollapsed: Bool = true   // For userResponse - collapsed by default
    var responseLatencyMs: Int?    // Time from question end to completed answer

    enum MessageType {
        case question      // Interviewer question
        case answer        // Generated AI answer
        case userResponse  // User's spoken response (detected as "answer")
        case followUp      // Follow-up answer
        case status        // System status message
        case screenshot    // Screenshot thumbnail in timeline
        case codingTask    // Pinned coding task solution (full width, larger)
    }

    init(type: MessageType, content: String, topic: String? = nil, screenshotId: UUID? = nil, audioSource: AudioSource? = nil, responseLatencyMs: Int? = nil) {
        self.id = UUID()
        self.timestamp = Date()
        self.type = type
        self.content = content
        self.topic = topic
        self.screenshotId = screenshotId
        self.audioSource = audioSource
        self.isCollapsed = (type == .userResponse)  // Collapsed by default for user responses
        self.responseLatencyMs = responseLatencyMs
    }

    var isQuestion: Bool {
        type == .question
    }

    var isAnswer: Bool {
        type == .answer || type == .followUp
    }

    var displayTime: String {
        let formatter = DateFormatter()
        formatter.timeStyle = .medium
        return formatter.string(from: timestamp)
    }

    var displayLatency: String? {
        guard let ms = responseLatencyMs else { return nil }
        if ms >= 1000 {
            let seconds = Double(ms) / 1000.0
            return String(format: "%.1fs", seconds)
        } else {
            return "\(ms)ms"
        }
    }
}

File: Domain/Model/InterviewMessage.cs
using System;

namespace InterviewMasterApp.Domain.Model
{
    /// <summary>
    /// Audio source for speaker identification.
    /// Migrated from Swift: AudioSource enum
    /// </summary>
    public enum AudioSource
    {
        Microphone,     // Your voice (you speaking)
        SystemAudio     // Interviewer voice (Zoom/Teams output)
    }

    /// <summary>
    /// Represents a message in the interview timeline.
    /// Migrated from Swift: InterviewMessage.swift
    /// </summary>
    public class InterviewMessage
    {
        public enum MessageType
        {
            Question,       // Interviewer question
            Answer,         // Generated AI answer
            UserResponse,   // User's spoken response (detected as "answer")
            FollowUp,       // Follow-up answer
            Status,         // System status message
            Screenshot,     // Screenshot thumbnail in timeline
            CodingTask      // Pinned coding task solution (full width, larger)
        }

        public Guid Id { get; }
        public DateTime Timestamp { get; }
        public MessageType Type { get; }
        public string Content { get; }
        public string? Topic { get; }
        public Guid? ScreenshotId { get; set; }
        public AudioSource? AudioSource { get; set; }
        public bool IsCollapsed { get; set; }
        public int? ResponseLatencyMs { get; set; }

        public InterviewMessage(
            MessageType type,
            string content,
            string? topic = null,
            Guid? screenshotId = null,
            AudioSource? audioSource = null,
            int? responseLatencyMs = null)
        {
            Id = Guid.NewGuid();
            Timestamp = DateTime.UtcNow;
            Type = type;
            Content = content ?? throw new ArgumentNullException(nameof(content));
            Topic = topic;
            ScreenshotId = screenshotId;
            AudioSource = audioSource;
            IsCollapsed = (type == MessageType.UserResponse);  // Collapsed by default for user responses
            ResponseLatencyMs = responseLatencyMs;
        }

        // MARK: - Computed Properties

        public bool IsQuestion => Type == MessageType.Question;

        public bool IsAnswer => Type == MessageType.Answer || Type == MessageType.FollowUp;

        public string DisplayTime => Timestamp.ToString("h:mm:ss tt");

        public string? DisplayLatency
        {
            get
            {
                if (ResponseLatencyMs == null)
                    return null;

                if (ResponseLatencyMs >= 1000)
                {
                    var seconds = (double)ResponseLatencyMs / 1000.0;
                    return string.Format("{0:F1}s", seconds);
                }
                else
                {
                    return $"{ResponseLatencyMs}ms";
                }
            }
        }

        public override bool Equals(object? obj)
        {
            return obj is InterviewMessage other && Id == other.Id;
        }

        public override int GetHashCode()
        {
            return Id.GetHashCode();
        }

        public override string ToString()
        {
            var preview = Content.Length > 50 ? Content.Substring(0, 50) + "..." : Content;
            return $"InterviewMessage(Type={Type}, Topic={Topic}, Content='{preview}')";
        }
    }
}
